eth/gasprice/gasprice.go

```diff
@@ -64,82 +71,101 @@ func NewOracle(backend ethapi.Backend, params Config) *Oracle {
 		percent = 100
 	}
 	return &Oracle{
-		backend:     backend,
-		lastPrice:   params.Default,
-		checkBlocks: blocks,
-		maxEmpty:    blocks / 2,
-		maxBlocks:   blocks * 5,
-		percentile:  percent,
+		backend:      backend,
+		defaultPrice: params.Default,
+		lastPrice:    params.Default,
+		checkBlocks:  blocks,
+		maxInvalid:   blocks / 2,
+		maxBlocks:    blocks * 5,
+		sampleNumber: sampleNumber,
+		percentile:   percent,
 	}
 }
 
-// SuggestPrice returns the recommended gas price.
-func (gpo *Oracle) SuggestPrice(ctx context.Context) (*big.Int, error) {
+// NewFullOracle returns a gasprice oracle for full node which has
+// avaiblable recent blocks in local db. FullOracle has higher
+// recommendation accuracy.
+func NewFullOracle(backend ethapi.Backend, params Config) *Oracle {
+	return newOracle(backend, fullSampleNumber, params)
+}
+
+// NewLightOracle returns a gasprice oracle for light client which doesn't
+// has recent block locally. LightOracle is much cheaper than FullOracle
+// however the corresponding recommendation accuracy is lower.
+func NewLightOracle(backend ethapi.Backend, params Config) *Oracle {
+	return newOracle(backend, lightSampleNumber, params)
+}
+
+// getLatest returns a recommended gas price which is suggested last time
+// but still suitable now.
+func (gpo *Oracle) getLatest(headHash common.Hash) *big.Int {
 	gpo.cacheLock.RLock()
 	lastHead := gpo.lastHead
 	lastPrice := gpo.lastPrice
 	gpo.cacheLock.RUnlock()
+	if headHash == lastHead {
```

Comment:
To me it looks a bit strange to have this check here since you still have to do another check for the return value anyway. Honestly I am not sure this function is necessary at all.